src/main/java/com/redhat/sast/api/service/AbstractBatchService.java

```diff
@@ -0,0 +1,222 @@
+     */
+    protected boolean canBeCancelled(JobStatus status) {
+        return status == JobStatus.RUNNING || status == JobStatus.SCHEDULED || status == JobStatus.PENDING;
+    }
+
+    /**
+     * Converts a JobBatch entity to a response DTO.
+     */
+    protected JobBatchResponseDto convertToResponseDto(@Nonnull JobBatch batch) {
+        JobBatchResponseDto dto = new JobBatchResponseDto();
+        dto.setBatchId(batch.getId());
+        dto.setBatchGoogleSheetUrl(batch.getBatchGoogleSheetUrl());
+        dto.setSubmittedBy(batch.getSubmittedBy());
+        dto.setSubmittedAt(batch.getSubmittedAt());
+        dto.setStatus(batch.getStatus());
+        dto.setTotalJobs(batch.getTotalJobs());
+        dto.setCompletedJobs(batch.getCompletedJobs());
+        dto.setFailedJobs(batch.getFailedJobs());
+        return dto;
+    }
+
+    /**
+     * Core job processing loop used by both SAST AI and MLOps pipelines.
+     * Template method that delegates pipeline-specific operations to abstract methods.
+     */
+    protected <T> void processJobsTemplate(
+            Long batchId,
+            List<T> jobDtos,
+            String pipelineName,
+            JobProcessor<T> processor) {
+        final AtomicInteger completedCount = new AtomicInteger(0);
+        final AtomicInteger failedCount = new AtomicInteger(0);
+
+        LOGGER.debug("{} Batch {}: Starting sequential processing of {} jobs", pipelineName, batchId, jobDtos.size());
+
+        for (int i = 0; i < jobDtos.size(); i++) {
+            T jobDto = jobDtos.get(i);
+            Long jobId = null;
+
+            try {
+                String packageNvr = processor.getPackageNvr(jobDto);
+                LOGGER.debug(
+                        "{} Batch {}: Processing job {}/{} (Package: {})",
+                        pipelineName,
+                        batchId,
+                        i + 1,
+                        jobDtos.size(),
+                        packageNvr);
+
+                final Job createdJob = processor.createJob(jobDto);
+                jobId = createdJob.getId();
+                final List<Param> pipelineParams = processor.extractPipelineParams(createdJob);
+                final String llmSecretName = (createdJob.getJobSettings() != null)
+                        ? createdJob.getJobSettings().getSecretName()
+                        : "sast-ai-default-llm-creds";
+
+                batchProcessingService.associateJobToBatchInNewTransaction(jobId, batchId);
+
+                LOGGER.debug(
+                        "{} Batch {}: Starting pipeline for job {} (Package: {})",
+                        pipelineName,
+                        batchId,
+                        jobId,
+                        packageNvr);
+
+                platformService.startPipeline(jobId, processor.getPipelineName(), pipelineParams, llmSecretName);
+
+                boolean jobCompleted = batchProcessingService.waitForJobCompletion(jobId);
```

Comment:
This is a blocking method call. It will easily timeout our http request for the serving user. Think about a execution of a large report with 1000 sast errors. Pipeline will be in `RUNNING` state for 30 mins. In that case, we check every 5 secs.